### DISCUSSION TITLE
Module Federation - increase in bundle size

### DISCUSSION THREAD

**Discussion by XiaofengXie16** [webpack/webpack#13382] [Q&A] [answered]
When I started to use module federation, I thought the overhead would just be the remoteEntry file (which is about 100 b) and some small fixed overhead  for creating the container. In reality, it seems like the bundle size grows as you add more components to the `exposed`  entry. For my use case, I exposed the entire app and the bundle size grows 10+KB after gzipped(40MB before gzipped).

 I can see the same trend  in the module-federation example. https://github.com/module-federation/module-federation-examples)

It would be great if someone can validate my observation here and explain to me what exactly happen behind the scene. Thank you in advance!

**Comment by alexander-akait** [comment]
Because you need some more runtime code to using module federation

**Comment by sokra** [comment, marked as answer]
Exposing modules will no longer allow to tree shake the exposed modules, since it's not known how these modules are used by remote modules.

It will also put exposed modules into separate chunks to load them only when used.

**Reply by XiaofengXie16** [reply]
Thank you for your response @sokra . Just want to get a bit more clarification on this.

I have an app shell and app A . App A is deployed as a standalone application but also exposes itself **entirely** as a remote . So users can access the standalone version of app A or via the shell.

Based on your answer, it seems to me that the standalone version will also lose tree-shaking capabilities because the entire app A is now an exposed modules . Do I understand it correctly?

**Reply by sokra** [reply]
> Based on your answer, it seems to me that the standalone version will also lose tree-shaking capabilities because the entire app A is now an exposed modules . Do I understand it correctly?

It depends on whether they use single runtime or multiple runtimes. Modules are optimized per runtime (in production mode), so if you don't specify `optimization.runtimeChunk: "single"` each entrypoint resp the MF entrypoint will have different modules.

Also make sure to measure correctly. Measure what the browser loads and don't measure the size of the output directory.

**Reply by gzaripov** [reply]
@sokra I found that remote entry file contains about 4-5KB of code that can be shared. Can we share it between remote entries? If you have a lot of microfrontends you can save dozens of kilobytes just reusing that runtime code.
